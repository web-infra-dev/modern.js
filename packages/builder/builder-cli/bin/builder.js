#!/usr/bin/env node
const { logger } = require('@modern-js/utils/logger');
const { run } = require('../dist');

async function main() {
  try {
    await run();
  } catch (err) {
    logger.error(err);
  }
}

main();
