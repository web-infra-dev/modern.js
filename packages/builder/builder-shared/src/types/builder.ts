import type { BuilderContext } from './context';
import type { PluginStore } from './plugin';
import type { BuilderProvider, ProviderInstance } from './provider';

export type BuilderTarget =
  | 'web'
  | 'node'
  | 'modern-web'
  | 'web-worker'
  | 'service-worker';

export type BuilderEntry = Record<string, string | string[]>;

export type BuilderMode = 'development' | 'production';

export type CreateBuilderOptions = {
  /** The root path of current project. */
  cwd?: string;
  /** The entry points object. */
  entry?: BuilderEntry;
  /** Type of build target. */
  target?: BuilderTarget | BuilderTarget[];
  /** Framework name, such as 'modern.js' */
  framework?: string;
  /** Absolute path of framework config file. */
  configPath?: string | null;
};

export type BuilderInstance<P extends BuilderProvider = BuilderProvider> = {
  context: BuilderContext;

  addPlugins: PluginStore['addPlugins'];
  removePlugins: PluginStore['removePlugins'];
  isPluginExists: PluginStore['isPluginExists'];

  serve: ProviderInstance['serve'];
  build: ProviderInstance['build'];
  initConfigs: ProviderInstance['initConfigs'];
  inspectConfig: ProviderInstance['inspectConfig'];
  createCompiler: ProviderInstance['createCompiler'];
  startDevServer: ProviderInstance['startDevServer'];

  getHTMLPaths: Awaited<ReturnType<P>>['pluginAPI']['getHTMLPaths'];
  getBuilderConfig: Awaited<ReturnType<P>>['pluginAPI']['getBuilderConfig'];
  getNormalizedConfig: Awaited<
    ReturnType<P>
  >['pluginAPI']['getNormalizedConfig'];

  onBeforeBuild: Awaited<ReturnType<P>>['pluginAPI']['onBeforeBuild'];
  onBeforeCreateCompiler: Awaited<
    ReturnType<P>
  >['pluginAPI']['onBeforeCreateCompiler'];
  onBeforeStartDevServer: Awaited<
    ReturnType<P>
  >['pluginAPI']['onBeforeStartDevServer'];
  onAfterBuild: Awaited<ReturnType<P>>['pluginAPI']['onAfterBuild'];
  onAfterCreateCompiler: Awaited<
    ReturnType<P>
  >['pluginAPI']['onAfterCreateCompiler'];
  onAfterStartDevServer: Awaited<
    ReturnType<P>
  >['pluginAPI']['onAfterStartDevServer'];
  onDevCompileDone: Awaited<ReturnType<P>>['pluginAPI']['onDevCompileDone'];
  onExit: Awaited<ReturnType<P>>['pluginAPI']['onExit'];
};
