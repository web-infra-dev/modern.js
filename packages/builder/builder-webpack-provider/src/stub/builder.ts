import type * as playwright from '@modern-js/e2e/playwright';
import { getTemplatePath } from '@modern-js/utils';
import _ from '@modern-js/utils/lodash';
import {
  createPluginStore,
  createPublicContext,
  applyDefaultBuilderOptions,
  type PluginStore,
  type BuildOptions,
  type CreateBuilderOptions,
} from '@modern-js/builder-shared';
import assert from 'assert';
import { PathLike } from 'fs';
import {
  applyBasicPlugins,
  applyDefaultPlugins,
  applyMinimalPlugins,
} from '../shared/plugin';
import { URL } from 'url';
import { Hooks } from '../core/initHooks';
import type { BuilderConfig, BuilderPlugin, Context } from '../types';
import { STUB_BUILDER_PLUGIN_BUILTIN } from './constants';
import { createStubContext } from './context';
import { matchLoader, globContentJSON, filenameToGlobExpr } from './utils';

export interface OptionsPluginsItem {
  builtin?: boolean | 'default' | 'minimal' | 'basic';
  additional?: BuilderPlugin[];
}

export interface StubBuilderOptions extends CreateBuilderOptions {
  context?: Context;
  /**
   * Setup builtin plugins and add custom plugins.
   * Automatically add builtin plugins by `process.env.STUB_BUILDER_PLUGIN_BUILTIN`.
   */
  plugins?: OptionsPluginsItem | OptionsPluginsItem[keyof OptionsPluginsItem];
  /**
   * Whether to run webpack build. By default it will be `false` and skip webpack building.
   * Set a string value to specify the `output.distPath` config.
   */
  webpack?: boolean | string;
  buildOptions?: BuildOptions;
  builderConfig?: BuilderConfig;
}

export type HookApi = {
  [key in keyof Hooks]: Parameters<Parameters<Hooks[key]['tap']>[0]>;
};

export interface ServeDestOptions {
  hangOn?: playwright.TestType<any, any> | boolean;
}

export function normalizeStubPluginOptions(
  options?: StubBuilderOptions['plugins'],
) {
  const opt: Required<OptionsPluginsItem> = {
    builtin: STUB_BUILDER_PLUGIN_BUILTIN ?? false,
    additional: [],
  };
  // normalize options
  if (Array.isArray(options)) {
    opt.additional = options;
  } else if (typeof options === 'object') {
    Object.assign(opt, options);
  } else if (typeof options === 'string' || typeof options === 'boolean') {
    opt.builtin = options;
  }
  return opt;
}

export async function applyPluginOptions(
  pluginStore: PluginStore,
  options?: StubBuilderOptions['plugins'],
) {
  const opt = normalizeStubPluginOptions(options);
  // apply plugins
  if (opt.builtin === true || opt.builtin === 'minimal') {
    pluginStore.addPlugins(await applyMinimalPlugins());
  } else if (opt.builtin === 'basic') {
    pluginStore.addPlugins(await applyBasicPlugins());
  } else if (opt.builtin === 'default') {
    pluginStore.addPlugins(await applyDefaultPlugins());
  }
  pluginStore.addPlugins(opt.additional);
}

/**
 * Create stub builder for testing.
 * Some behaviors will be different to common `createBuilder`.
 */
export async function createStubBuilder(options?: StubBuilderOptions) {
  // init primary builder.
  const builderOptions = applyDefaultBuilderOptions(
    options,
  ) as Required<StubBuilderOptions>;
  // apply webpack option.
  if (options?.webpack) {
    const distPath =
      typeof options.webpack === 'string'
        ? options.webpack
        : getTemplatePath('modern-js/stub-builder/dist');
    _.set(builderOptions.builderConfig, 'output.distPath', distPath);
  }
  // init context.
  const context = createStubContext(builderOptions);
  // merge user context.
  options?.context && _.merge(context, options.context);

  const publicContext = createPublicContext(context);
  const pluginStore = createPluginStore();

  // add builtin and custom plugins by `options.plugins`.
  await applyPluginOptions(pluginStore, options?.plugins);

  // tap on each hook and cache the args.
  const resolvedHooks: Record<string, any> = {};
  _.each(context.hooks, ({ tap }, name) => {
    tap((...args) => {
      resolvedHooks[name] = args;
    });
  });

  /**
   * Run build and caching args of all hooks.
   * It will run actual build only first time,
   * and always return cached result until {@link reset}.
   */
  const build = _.memoize(async () => {
    const { build: buildImpl, webpackBuild } = await import('../core/build');
    const executeBuild = options?.webpack ? webpackBuild : undefined;
    await buildImpl(
      { context, pluginStore, builderOptions },
      options?.buildOptions,
      executeBuild,
    );
    return { resolvedHooks: { ...resolvedHooks } };
  });

  /**
   * Unwrap args of hook.
   *
   * The `unwrap` means it will try to run {@link build} and run some asserts before return.
   */
  const unwrapHook = async <T extends keyof HookApi>(
    hook: T,
  ): Promise<HookApi[T]> => (await build()).resolvedHooks[hook];

  /** Unwrap webpack configs. */
  const unwrapWebpackConfigs = async () => {
    const [{ bundlerConfigs }] = await unwrapHook('onBeforeBuildHook');
    return bundlerConfigs;
  };

  /** Unwrap webpack config, it will ensure there's only one config object. */
  const unwrapWebpackConfig = async () => {
    const webpackConfigs = await unwrapWebpackConfigs();
    assert(webpackConfigs.length === 1);
    return webpackConfigs[0];
  };

  /** Unwrap webpack compiler instance. */
  const unwrapWebpackCompiler = async () => {
    const [{ compiler }] = await unwrapHook('onAfterCreateCompilerHooks');
    return compiler;
  };

  /** Serialize content of output files into JSON object. */
  const unwrapOutputJSON = async (
    paths: PathLike | PathLike[] = context.distPath,
    isRelative = false,
    maxSize = 4096,
  ) => {
    if (Array.isArray(paths) && isRelative) {
      throw new Error('`isRelative` is not supported for multiple paths.');
    }
    await build();
    const _paths = _(paths)
      .castArray()
      .map(filenameToGlobExpr)
      .map(String)
      .value();
    return globContentJSON(_paths, { absolute: !isRelative, maxSize });
  };

  /** Read output file content. */
  const readOutputFile = async (
    filename: string,
  ): Promise<string | undefined> => {
    const compiler = await unwrapWebpackCompiler();
    return new Promise((resolve, reject) => {
      compiler.outputFileSystem.readFile(filename, (err, data) =>
        err ? reject(err) : resolve(data?.toString('utf-8')),
      );
    });
  };

  /** Unwrap output file content, will ensure it is not empty. */
  const unwrapOutputFile = async (filename: string) => {
    await build();
    const content = await readOutputFile(filename);
    assert(content);
    return content;
  };

  /**
   * Run build and serve on `distPath`.
   * If `hangOn` is true, it will block running and let you debug the server.
   * Set `hangOn` to a {@link playwright.TestType} to automatically unset timeout of test.
   *
   * @example
   * test('should work', async () => {
   *   const builder = await createStubBuilder();
   *   const { baseurl } = builder.buildAndServe({ hangOn: test });
   *   page.goto(`${baseurl}/index.html`);
   * });
   */
  const buildAndServe = async (options?: ServeDestOptions) => {
    const [{ runStaticServer }] = await Promise.all([
      import('@modern-js/e2e'),
      build(),
    ]);
    const { port } = await runStaticServer(context.distPath);
    if (options?.hangOn) {
      // eslint-disable-next-line no-console
      console.log(
        `Successfully build, and hang on running server: http://localhost:${port}`,
      );
      typeof options.hangOn !== 'boolean' && options.hangOn.setTimeout(0);
      await new Promise(() => null);
    }
    // TODO: get real `htmlRoot` & `homeUrl` from compiler.
    const baseUrl = new URL(`http://localhost:${port}`);
    const htmlRoot = new URL('html/', baseUrl);
    const homeUrl = new URL('index/index.html', htmlRoot);
    return {
      baseUrl,
      htmlRoot,
      homeUrl,
      port,
    };
  };

  /** Match webpack plugin by constructor name. */
  const matchWebpackPlugin = async (pluginName: string) => {
    const config = await unwrapWebpackConfig();
    const result = config.plugins?.filter(
      item => item.constructor.name === pluginName,
    );
    if (Array.isArray(result)) {
      assert(result.length <= 1);
      return result[0] || null;
    } else {
      return result || null;
    }
  };

  /** Check if a file handled by specific loader. */
  const matchWebpackLoader = async (loader: string, testFile: string) =>
    matchLoader({ config: await unwrapWebpackConfig(), loader, testFile });

  /** Reset builder. @see {@link build} */
  const reset = () => {
    build.cache.clear!();
  };

  return {
    ...pluginStore,
    build,
    hooks: context.hooks,
    context,
    publicContext,
    unwrapHook,
    unwrapWebpackConfigs,
    unwrapWebpackConfig,
    unwrapWebpackCompiler,
    unwrapOutputJSON,
    unwrapOutputFile,
    readOutputFile,
    buildAndServe,
    matchWebpackPlugin,
    matchWebpackLoader,
    reset,
  };
}
