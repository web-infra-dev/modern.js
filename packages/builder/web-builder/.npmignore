.DS_Store

.pnp
.pnp.js
.env.local
.env.*.local
.history
.rts*
*.log*
*.pid
*.pid.*
*.report
*.lcov
lib-cov

node_modules/
.npm
.lock-wscript
.yarn-integrity
.node_repl_history
.nyc_output
*.tsbuildinfo
.eslintcache
.sonarlint

coverage/
release/
tests/

.vscode/**/*
!.vscode/settings.json
!.vscode/extensions.json
.idea/

src/

modern.config.js
modern.config.ts
jest.config.js
.eslintrc.js
.eslintrc
tsconfig.json

.wireit/
