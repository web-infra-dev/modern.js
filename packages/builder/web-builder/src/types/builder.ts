import type { BuilderConfig } from './config';

export type BuilderTarget = 'web' | 'node' | 'modern-web';

export type BuilderOptions = {
  cwd?: string;
  target?: BuilderTarget | BuilderTarget[];
  configPath?: string | null;
  builderConfig?: BuilderConfig;
  framework?: string;
};
