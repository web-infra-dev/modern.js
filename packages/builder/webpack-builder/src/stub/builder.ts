import type * as playwright from '@modern-js/e2e/playwright';
import _ from '@modern-js/utils/lodash';
import assert from 'assert';
import { PathLike } from 'fs';
import { DirectoryJSON, Volume } from 'memfs/lib/volume';
import path from 'path';
import { URL } from 'url';
import { webpackBuild } from '../core/build';
import { addDefaultPlugins, createPrimaryBuilder } from '../core/createBuilder';
import { Hooks } from '../core/createHook';
import { matchLoader, mergeBuilderOptions } from '../shared';
import type {
  BuilderOptions,
  BuilderPlugin,
  Context,
  PluginStore,
} from '../types';
import { STUB_BUILDER_PLUGIN_BUILTIN } from './constants';
import { createStubContext } from './context';
import { globContentJSON, filenameToGlobExpr } from './utils';

export interface OptionsPluginsItem {
  builtin?: boolean | 'default' | 'minimal';
  additional?: BuilderPlugin[];
}

export interface StubBuilderOptions extends BuilderOptions {
  context?: Context;
  /**
   * Setup builtin plugins and add custom plugins.
   * Automatically add builtin plugins by `process.env.STUB_BUILDER_PLUGIN_BUILTIN`.
   */
  plugins?: OptionsPluginsItem | OptionsPluginsItem[keyof OptionsPluginsItem];
  /** Whether to run webpack build. By default it will be `false` and skip webpack building. */
  webpack?: boolean | 'in-memory';
}

export type HookApi = {
  [key in keyof Hooks]: Parameters<Parameters<Hooks[key]['tap']>[0]>;
};

export interface ServeDestOptions {
  hangOn?: playwright.TestType<any, any> | boolean;
}

export function normalizeStubPluginOptions(
  options?: StubBuilderOptions['plugins'],
) {
  const opt: Required<OptionsPluginsItem> = {
    builtin: STUB_BUILDER_PLUGIN_BUILTIN ?? false,
    additional: [],
  };
  // normalize options
  if (Array.isArray(options)) {
    opt.additional = options;
  } else if (typeof options === 'object') {
    Object.assign(opt, options);
  } else if (typeof options === 'string' || typeof options === 'boolean') {
    opt.builtin = options;
  }
  return opt;
}

export async function applyPluginOptions(
  pluginStore: PluginStore,
  options?: StubBuilderOptions['plugins'],
) {
  const opt = normalizeStubPluginOptions(options);
  // apply plugins
  if (opt.builtin === true || opt.builtin === 'minimal') {
    // TODO: load minimal plugins
  } else if (opt.builtin === 'default') {
    await addDefaultPlugins(pluginStore);
  }
  pluginStore.addPlugins(opt.additional);
}

/**
 * Create stub builder for testing.
 * Some behaviors will be different to common `createBuilder`.
 */
export async function createStubBuilder(options?: StubBuilderOptions) {
  // init primary builder.
  const builderOptions = mergeBuilderOptions(
    options,
  ) as Required<StubBuilderOptions>;
  const context = createStubContext(builderOptions);
  options?.context && _.merge(context, options.context);
  const {
    pluginStore,
    publicContext,
    build: buildImpl,
  } = createPrimaryBuilder(builderOptions, context);
  await applyPluginOptions(pluginStore, options?.plugins);

  // replace outputFileSystem of Webpack.
  let memfsVolume: Volume | undefined;
  context.hooks.onAfterCreateCompilerHooks.tap(async ({ compiler }) => {
    if (options?.webpack === 'in-memory') {
      const { createFsFromVolume, Volume } = await import('memfs');
      const vol = new Volume();
      const ofs = createFsFromVolume(vol);
      memfsVolume = vol;
      compiler.outputFileSystem = ofs;
    }
  });

  // tap on each hook and cache the args.
  const resolvedHooks: Record<string, any> = {};
  _.each(context.hooks, ({ tap }, name) => {
    tap((...args) => {
      resolvedHooks[name] = args;
    });
  });

  /**
   * Run build and caching args of all hooks.
   * It will run actual build only first time,
   * and always return cached result until {@link reset}.
   */
  const build = _.memoize(async () => {
    const executeBuild = options?.webpack ? webpackBuild : undefined;
    await buildImpl(executeBuild);
    return { resolvedHooks: { ...resolvedHooks } };
  });

  /**
   * Unwrap args of hook.
   *
   * The `unwrap` means it will try to run {@link build} and run some asserts before return.
   */
  const unwrapHook = async <T extends keyof HookApi>(
    hook: T,
  ): Promise<HookApi[T]> => (await build()).resolvedHooks[hook];

  /** Unwrap webpack configs. */
  const unwrapWebpackConfigs = async () => {
    const [{ webpackConfigs }] = await unwrapHook('onBeforeBuildHook');
    return webpackConfigs;
  };

  /** Unwrap webpack config, it will ensure there's only one config object. */
  const unwrapWebpackConfig = async () => {
    const webpackConfigs = await unwrapWebpackConfigs();
    assert(webpackConfigs.length === 1);
    return webpackConfigs[0];
  };

  /** Unwrap webpack compiler instance. */
  const unwrapWebpackCompiler = async () => {
    const [{ compiler }] = await unwrapHook('onAfterCreateCompilerHooks');
    return compiler;
  };

  /** Unwrap outputFileSystem of webpack and ensure it is {@link Volume}. */
  const unwrapOutputVolume = async () => {
    await build();
    assert(memfsVolume);
    return memfsVolume;
  };

  /** Serialize content of output files into JSON object. */
  const unwrapOutputJSON = async (
    paths: PathLike | PathLike[] = context.distPath,
    isRelative = false,
    maxSize = 4096,
  ): Promise<DirectoryJSON> => {
    if (Array.isArray(paths) && isRelative) {
      throw new Error('`isRelative` is not supported for multiple paths.');
    }
    await build();
    if (memfsVolume) {
      // avoid memfs remove drive letter on windows, refer to https://github.com/streamich/memfs/issues/316.
      if (!isRelative && process.platform === 'win32') {
        const ret = memfsVolume.toJSON(paths, undefined, true);
        return _.mapKeys(ret, (_v, k) => path.join(paths as string, k));
      } else {
        const ret = memfsVolume.toJSON(paths, undefined, isRelative);
        return ret;
      }
    } else {
      const _paths = _(paths)
        .castArray()
        .map(filenameToGlobExpr)
        .map(String)
        .value();
      return globContentJSON(_paths, { absolute: !isRelative, maxSize });
    }
  };

  /** Read output file content. */
  const readOutputFile = async (
    filename: string,
  ): Promise<string | undefined> => {
    const compiler = await unwrapWebpackCompiler();
    return new Promise((resolve, reject) => {
      compiler.outputFileSystem.readFile(filename, (err, data) =>
        err ? reject(err) : resolve(data?.toString('utf-8')),
      );
    });
  };

  /** Unwrap output file content, will ensure it is not empty. */
  const unwrapOutputFile = async (filename: string) => {
    await build();
    const content = await readOutputFile(filename);
    assert(content);
    return content;
  };

  /**
   * Run build and serve on `distPath`.
   * If `hangOn` is true, it will block running and let you debug the server.
   * Set `hangOn` to a {@link playwright.TestType} to automatically unset timeout of test.
   *
   * @example
   * test('should work', async () => {
   *   const builder = await createStubBuilder();
   *   const { baseurl } = builder.buildAndServe({ hangOn: test });
   *   page.goto(`${baseurl}/index.html`);
   * });
   */
  const buildAndServe = async (options?: ServeDestOptions) => {
    const [{ runStaticServer }] = await Promise.all([
      import('@modern-js/e2e'),
      build(),
    ]);
    const { port } = await runStaticServer(context.distPath, {
      volume: memfsVolume,
    });
    if (options?.hangOn) {
      // eslint-disable-next-line no-console
      console.log(
        `Successfully build, and hang on running server: http://localhost:${port}`,
      );
      typeof options.hangOn !== 'boolean' && options.hangOn.setTimeout(0);
      await new Promise(() => null);
    }
    // TODO: get real `htmlRoot` & `homeUrl` from compiler.
    const baseUrl = new URL(`http://localhost:${port}`);
    const htmlRoot = new URL('html/', baseUrl);
    const homeUrl = new URL('index/index.html', htmlRoot);
    return {
      baseUrl,
      htmlRoot,
      homeUrl,
      volume: memfsVolume,
      port,
    };
  };

  /** Match webpack plugin by constructor name. */
  const matchWebpackPlugin = async (pluginName: string) => {
    const config = await unwrapWebpackConfig();
    const result = config.plugins?.filter(
      item => item.constructor.name === pluginName,
    );
    if (Array.isArray(result)) {
      assert(result.length <= 1);
      return result[0] || null;
    } else {
      return result || null;
    }
  };

  /** Check if a file handled by specific loader. */
  const matchWebpackLoader = async (loader: string, testFile: string) =>
    matchLoader({ config: await unwrapWebpackConfig(), loader, testFile });

  /** Reset builder. @see {@link build} */
  const reset = () => {
    build.cache.clear!();
  };

  return {
    ...pluginStore,
    build,
    hooks: context.hooks,
    context,
    publicContext,
    unwrapHook,
    unwrapWebpackConfigs,
    unwrapWebpackConfig,
    unwrapWebpackCompiler,
    unwrapOutputVolume,
    unwrapOutputJSON,
    unwrapOutputFile,
    readOutputFile,
    buildAndServe,
    matchWebpackPlugin,
    matchWebpackLoader,
    reset,
  };
}
