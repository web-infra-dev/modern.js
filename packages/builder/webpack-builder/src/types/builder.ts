import type { EntryObject } from 'webpack';
import type { BuilderConfig } from './config';

export type BuilderTarget = 'web' | 'node' | 'modern-web';

export type BuilderMode = 'development' | 'production';

export type BuilderOptions = {
  /** The root path of current project. */
  cwd?: string;
  /** The entry points object. */
  entry?: EntryObject;
  /** Type of build target. */
  target?: BuilderTarget | BuilderTarget[];
  /** Framework name, such as 'modern.js' */
  framework?: string;
  /** Absolute path of framework config file. */
  configPath?: string | null;
  /** Builder config object. */
  builderConfig?: BuilderConfig;
};

export type InspectOptions = {
  env?: BuilderMode;
  verbose?: boolean;
  outputPath?: string;
  writeToDisk?: boolean;
};
