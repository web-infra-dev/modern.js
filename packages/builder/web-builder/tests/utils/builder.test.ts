import { expect, it, describe } from 'vitest';
import { createStubBuilder } from './builder';

describe('tests/stub-builder', () => {
  it('should catch trace stacks when hooks unresolved', async () => {
    const builder = createStubBuilder();
    await builder.unwrapWebpackConfig();
    await expect(builder.unwrapWebpackConfig()).rejects.toMatchSnapshot();
  });
});
