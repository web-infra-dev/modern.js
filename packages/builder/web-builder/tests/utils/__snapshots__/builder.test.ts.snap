// Vitest Snapshot v1

exports[`tests/stub-builder > should catch trace stacks when hooks unresolved 1`] = `Promise {}`;
